File: cli/narrator/chapter.py
def detect_chapter_title(page_text: str) -> str | None:
    lines = [l.strip() for l in page_text.strip().split('\n') if l.strip()]
    if len(lines) < 2:
        return None

    first = lines[0]
    words = first.split()

    # Must be short — chapter titles are rarely more than 8 words
    if not (1 <= len(words) <= 8):
        return None

    # Titles don't end with sentence punctuation
    if first.endswith(('.', ',', ':', ';', '?', '!')):
        return None

    # Body text after the title must be substantial
    body = ' '.join(lines[1:])
    if len(body.split()) < 15:
        return None

    return first


def build_chapter_map(pages: list[str]) -> dict[int, str]:
    chapter_map: dict[int, str] = {}
    for i, text in enumerate(pages):
        title = detect_chapter_title(text)
        if title:
            chapter_map[i] = title
    return chapter_map

File: pwa/src/lib/chapter.ts
export function detectChapterTitle(pageText: string): string | null {
  const lines = pageText
    .trim()
    .split('\n')
    .map((l) => l.trim())
    .filter(Boolean);

  if (lines.length < 2) return null;

  const first = lines[0];
  const words = first.split(/\s+/);

  if (words.length < 1 || words.length > 8) return null;
  if (/[.,;:?!]$/.test(first)) return null;

  const body = lines.slice(1).join(' ');
  if (body.split(/\s+/).length < 15) return null;

  return first;
}

export function buildChapterMap(pages: string[]): Record<number, string> {
  const map: Record<number, string> = {};
  for (let i = 0; i < pages.length; i++) {
    const title = detectChapterTitle(pages[i]);
    if (title) map[i] = title;
  }
  return map;
}
